# 01. Importing libraries 

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
path= r'C:\Users\Brooke\05-2022 Instacart Basket Analysis'

In [7]:
# Importing prepared orders data 
df_ords = pd.read_csv(os.path.join(path,'Data', 'Prepared Data', 'orders_wrangled.csv'))

In [11]:
# Importing products data 
df_prods = pd.read_csv(os.path.join(path,'Data', 'Original Data', 'products[1].csv'))

# Preparing data

In [12]:
# Create a dataframe
df_test = pd.DataFrame()

In [13]:
#create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [14]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [15]:
# Check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [16]:
#Changing mix column to string data type
df_test['mix'] = df_test['mix'].astype('str')

In [17]:
# identify missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [20]:
#creating missing values dataframe
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
df_prods.shape

(49693, 5)

In [23]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
# creating duplicates dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
# create dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [29]:
df_prods_clean_no_dups.shape

(49672, 5)

In [32]:
#exported data
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'products_checked.csv'))

In [31]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

# Maximum in the prices column looks to be far too high based on the minimum, mean, and median.

In [34]:
# checking for mixed-type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [35]:
# checking for missing data
df_ords.isnull().sum()

user_id                        0
order_id                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# The days_since_prior_order column has 206209 missing values.  This could be because there were observations in which there were no prior orders.

In [38]:
# add column for having prior orders or not
df_ords['has_prior_orders'] = df_ords['days_since_prior_order']>=0

In [39]:
df_ords.head()

user_id  order_id  order_number  orders_day_of_week  order_hour_of_day  \
0        1   2539329             1                   2                  8   
1        1   2398795             2                   3                  7   
2        1    473747             3                   3                 12   
3        1   2254736             4                   4                  7   
4        1    431534             5                   4                 15   

   days_since_prior_order  no_prior_orders  has_prior_orders  
0                     NaN            False             False  
1                    15.0             True              True  
2                    21.0             True              True  
3                    29.0             True              True  
4                    28.0             True              True

In [40]:
#correct error made in naming column incorrectly and needing to drop the inccorect column
df_ords_clean = df_ords.drop('no_prior_orders', 1)

C:\Users\Brooke\AppData\Local\Temp\ipykernel_31940\1318638294.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df_ords_clean = df_ords.drop('no_prior_orders', 1)


In [42]:
df_ords_clean.head()

user_id  order_id  order_number  orders_day_of_week  order_hour_of_day  \
0        1   2539329             1                   2                  8   
1        1   2398795             2                   3                  7   
2        1    473747             3                   3                 12   
3        1   2254736             4                   4                  7   
4        1    431534             5                   4                 15   

   days_since_prior_order  has_prior_orders  
0                     NaN             False  
1                    15.0              True  
2                    21.0              True  
3                    29.0              True  
4                    28.0              True

# I added a column to address whether or not that observation had prior orders.  The missing values would then be explained by the next column that states that there were no prior orders for that observation.

In [43]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [44]:
df_ords_dups

Empty DataFrame
Columns: [user_id, order_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, has_prior_orders]
Index: []

# There were no duplicated found.

In [45]:
#exported data
df_ords_clean.to_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_checked.csv'))